cpp/pagerank_module/algorithm/pagerank.hpp

```diff
@@ -0,0 +1,181 @@
+    }
+  }
+
+  /// Be careful and don't call AddAdjacentPair while you have reference to this
+  /// vector because the referenced vector could be resized (moved) which means
+  /// that the reference is going to become invalid.
+  ///
+  /// @return A reference to std::vector of adjecent node ids.
+  const auto &GetAdjacentNodes(T node_id) const { return list_[node_id]; }
+
+ private:
+  std::vector<std::vector<T>> list_;
+};
+
+}  // namespace pagerank_util
+
+namespace pagerank_alg {
+
+// Defines edge's from and to node index. Just an alias for user convenience.
+using EdgePair = std::pair<std::uint64_t, std::uint64_t>;
+
+/// A directed, unweighted graph.
+/// Self loops and multiple edges are allowed and they will affect the result.
+/// Node ids are integers from interval [0, number of nodes in graph - 1].
+/// Graph is allowed to be disconnected.
+class PageRankGraph {
+ public:
+  /// Creates graph with given number of nodes with node ids from interval
+  /// [0, number_of_nodes - 1] and with given edges between them.
+  /// Node ids describing edges have to be integers from
+  /// interval [0, number of nodes in graph - 1].
+  /// @param number_of_nodes -- number of nodes in graph
+  /// @param number_of_edges -- number of edges in graph
+  /// @param edges -- pairs (source, target) representing directed edges
+  PageRankGraph(std::uint64_t number_of_nodes, std::uint64_t number_of_edges, const std::vector<EdgePair> &edges);
+
+  /// @return -- number of nodes in graph
+  std::uint64_t GetNodeCount() const;
+
+  /// @return -- nubmer of edges in graph
+  std::uint64_t GetEdgeCount() const;
+
+  /// @return -- a reference to ordered ordered vector of edges
+  const std::vector<EdgePair> &GetOrderedEdges() const;
+
+  /// Returns out degree of node node_id
+  /// @param node_id -- node name
+  /// @return -- out degree of node node_id
+  std::uint64_t GetOutDegree(std::uint64_t node_id) const;
+
+ private:
+  /// node_count equals number of nodes in graph
+  std::uint64_t node_count_;
+  /// edge_count equals number of edges in graph
+  std::uint64_t edge_count_;
+  /// directed edges (source, target) (source -> target) ordered by target
+  std::vector<EdgePair> ordered_edges_;
+  /// out degree for each node in graph because it is required in calculating
+  /// PageRank
+  std::vector<std::uint64_t> out_degree_;
+};
+
+/// Calculates optimal borders for dividing edges in number_of_threads
+/// consecutive partitions (blocks) such that the maximal block size is minimal
+/// For example: if number_of_edges = 10 and number_of_threads = 3:
+/// optimal borders = {0, 3, 6, 10} so obtained blocks of edges
+/// are [0, 3>, [3, 6> and [6, 10>.
+///
+/// @param graph -- graph
+/// @param number_of_threads -- number of threads
+/// @param borders -- vector in which calculated optimal borders will be written
+void CalcualteOptimalBorders(const PageRankGraph &graph, int number_of_threads, std::vector<std::uint64_t> *borders);
```

Comment:
```suggestion
void CalculateOptimalBorders(const PageRankGraph &graph, int number_of_threads, std::vector<std::uint64_t> &borders);
```
So you basically declare one function here, and implement a completely different one in the cpp. :smile: